src/mito2/src/memtable/merge_tree/merger.rs

```diff
@@ -0,0 +1,633 @@
+
+pub trait Item: Clone {
+    type Key: Ord;
+
+    /// Remaining rows in item.
+    fn remaining(&self) -> usize;
+
+    /// Whether current item is exhausted.
+    fn is_empty(&self) -> bool;
+
+    /// The key range of item.
+    fn current_range(&self) -> Range<Self::Key>;
+
+    /// Searches given key in current item.
+    fn search_key(&self, key: &Self::Key) -> std::result::Result<usize, usize>;
+
+    /// Slice item.
+    fn slice(&self, range: Range<usize>) -> Self;
+}
+
+pub trait Node: Ord {
+    type Item: Item;
+
+    /// Returns current item of node and fetch next.
+    fn fetch_next(&mut self) -> Result<Self::Item>;
+
+    /// Returns true if current node is not exhausted.
+    fn is_valid(&self) -> bool;
+
+    /// Current item of node.
+    fn current_item(&self) -> &Self::Item;
+
+    /// Whether the other node is behind (exclusive) current node.
+    fn is_behind(&self, other: &Self) -> bool;
+
+    /// Skips first `num_to_skip` rows from node's current batch. If current batch is empty it fetches
+    /// next batch from the node.
+    ///
+    /// # Panics
+    /// If the node is EOF.
+    fn skip(&mut self, offset_to_skip: usize) -> Result<()>;
+}
+
+pub struct Merger<T, I> {
+    heap: BinaryHeap<T>,
+    current_item: Option<I>,
+}
+
+impl<T, I> Merger<T, I>
+where
+    I: Item + Debug,
+    T: Node<Item = I>,
+{
+    pub(crate) fn try_new(nodes: Vec<T>) -> Result<Self> {
+        let mut heap = BinaryHeap::with_capacity(nodes.len());
+        for node in nodes {
+            if node.is_valid() {
+                heap.push(node);
+            }
+        }
+        let mut merger = Merger {
+            heap,
+            current_item: None,
+        };
+        merger.next()?;
+        Ok(merger)
+    }
+
+    /// Returns true if current merger is still valid.
+    pub(crate) fn is_valid(&self) -> bool {
+        self.current_item.is_some()
+    }
+
+    /// Advances current merger to next item.
+    pub(crate) fn next(&mut self) -> Result<()> {
+        if let Some(mut top_node) = self.heap.pop() {
+            if let Some(next_node) = self.heap.peek() {
+                if next_node.is_behind(&top_node) {
+                    // does not overlap
+                    self.current_item = Some(top_node.fetch_next()?);
+                } else {
+                    let next_start = next_node.current_item().current_range().start;
+                    let res = match top_node.current_item().search_key(&next_start) {
+                        Ok(pos) => {
+                            // duplicate timestamp found, yield duplicated row in this item
+                            if pos == 0 {
+                                let to_yield = top_node.current_item().slice(0..1);
+                                top_node.skip(1)?;
+                                to_yield
+                            } else {
+                                let to_yield = top_node.current_item().slice(0..pos);
```

Comment:
Looks like we can always slice `0..pos+1`

Reply:
If we always slice `0..pos+1`, then it would break the sequence desc order contract.

Suppose current node has two rows: `(ts=0, seq=0), (ts=1, seq=1)` and next node has `(ts=1, seq=2)`, if we slice `0..pos+1` then `(ts=1, seq=1)` will be yielded along with `(ts=0, seq=0)`, not after `(ts=1, seq=2)`.